Vector should preserve warnings

hubertp opened the pull request:
### Pull Request Description

When Vector was created via a sequence literal, we simply dropped any associated any warnings associated with it.
Similarly doing any mapping on a Vector with warnings would create a new Vector with warnings now missing.

Rather than ducktaping every creation of `Array`/`Vector`/`whatever` so that they propagate warnings, decided to address the problem at the root. Array-like structures no longer remove warnings from values. Instead, whoever uses the values should ensure, if necessary, that they are gone. This was mostly problematic for polyglot calls.
Added a `WarningsLibrary` that exposes `hasWarnings`, `getWarnings` and `removeWarnings` to deal with warnings-related logic. This is primarily used in Vector which delegates all calls to the underlying storage but can be used in any collection.

Bonus points: warnings now survive multi-dimensional Vectors.

### Important note

This PR addresses:
- https://www.pivotaltracker.com/n/projects/2539304/stories/182221022
- https://www.pivotaltracker.com/n/projects/2539304/stories/182194556

(initially the solution was very limited as it only addressed the former).

While working on polyglot calls discovered by accident that `Warning.detach_selected_warnings` was buggy and not tested. This is now fixed.

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide build` and `./run ide watch`.

hubertp committed:
Sequence literal (Vector) should preserve warnings

When Vector was created via a sequence literal, we simply dropped any
associated any warnings associated with it.
This change propagates Warnings during the creation of the Vector.
Ideally, it would be sufficient to propagate warnings from the
individual elements to the underlying storage but doesn't go well with
`Vector.fromArray`.

hubertp committed:
update changelog

JaroslavTulach commented on test/Tests/src/Semantic/Warnings_Spec.enso:
```diff
@@ -157,4 +157,13 @@ spec = Test.group "Dataflow Warnings" <|
+        d = Warning.attach "d" c
+        c . should_equal [1, 2]
+        Warning.get_all c . map .value . should_equal ["b", "a"]
+        Warning.get_all d . map .value . should_equal ["d", "b", "a"]
```
Is the order really preserved? Does it have to be? Shouldn't the test just verify the right warnings are present, but ignore the order?

hubertp replied:
I thought the same but then found this: https://github.com/enso-org/enso/blob/develop/distribution/lib/Standard/Base/0.0.0-dev/src/Warning.enso#L30.
Apparently it is fixed to be reverse-chronological.

hubertp committed:
Array-like structures preserver warnings

Added a WarningsLibrary that exposes `hasWarnings` and `getWarnings`
messages. That way we can have a single storage that defines how to
extract warnings from an Array and the others just delegate to it.

This simplifies logic added to sequence literals to handle warnings.

radeusgd commented on engine/runtime/src/main/java/org/enso/interpreter/node/callable/SequenceLiteralNode.java:
```diff
@@ -35,9 +41,19 @@ public static SequenceLiteralNode build(ExpressionNode[] items) {
+    ArrayRope<Warning> warnings = new ArrayRope<>();
     for (int i = 0; i < items.length; i++) {
       itemValues[i] = items[i].executeGeneric(frame);
+      if (itemValues[i] instanceof WithWarnings) {
+        WithWarnings withWarnings = (WithWarnings) itemValues[i];
```
```suggestion
      if (itemValues[i] instanceof WithWarnings withWarnings) {
```

Any reason not to use the new Java features if we are already compiling with Frgaal?

radeusgd commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/ArrayRope.java:
```diff
@@ -106,6 +117,11 @@ public int size() {
 
+    @Override
+    public boolean isEmpty() {
+      return left.isEmpty() && right.isEmpty();
```
```suggestion
      return size() == 0;
```

Why not re-use the logic for computing the size which includes caching?

The current implementation will need to traverse the whole rope on _every_ call, whereas with caching we can respond immediately on subsequent calls. And we already have this logic implemented.

hubertp replied:
This logic is no longer needed, so removed. Also, it seemed that just finding the first non-empty segment was more efficient. But I agree that amortized cost would probably be better with calculating the size upfront.

radeusgd commented on test/Tests/src/Semantic/Warnings_Spec.enso:
```diff
@@ -157,4 +157,13 @@ spec = Test.group "Dataflow Warnings" <|
         result = x.is_static_nothing x
         result . should_equal False
 
+    Test.specify "should be allowed in Vector" <|
```
Can we also check if stuff like `Vector.map` will handle them correctly too? Or is this out of scope of this PR (fine if yes)?

I'd appreciate seeing something like:
```
v = 0.up_to 10 . to_vector
r = v.map (if x % 2 == 0 then 10*x else Warning.attach 100*x x)
r.should_equal [0, 1, 20, 3, 40, 5, 60, 7, 80, 9]
Warning.get_all r . map .value . should_contain_the_same_elements_as [100, 300, 500, 700, 900]
```

hubertp replied:
Unlikely this will work and on purpose. This is a subject of https://www.pivotaltracker.com/n/projects/2539304/stories/182194556. The fix for the latter might actually render this change obsolete.

radeusgd replied:
Ok, I hoped that maybe you managed to fix two birds with one ~~stone~~ _bugfix_.

But then, the latter may wait for its turn :)

hubertp replied:
It is now

hubertp committed:
Ensure polyglot method calls are warning-free

Since warnings are no longer automatically extracted from Array-like
structures, we delay the operation until an actual polyglot method call
is performed.

Discovered a bug in `Warning.detach_selected_warnings` which was missing
any usage or tests.

hubertp committed:
nits

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/error/WarningsLibrary.java:
```diff
@@ -0,0 +1,63 @@
+   * @return the associated warnings
+   */
+  @GenerateLibrary.Abstract(ifExported = {"hasWarnings"})
+  public Warning[] getWarnings(Object receiver, Node location) throws UnsupportedMessageException {
```
Wouldn't it be easier to return any polyglot object that `hasArrayElements`? That might save few conversions/wrappings later.

hubertp replied:
Can't do, it seems. Warnings need to be sorted later with per-creation timestamp

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/error/WithWarnings.java:
```diff
@@ -42,16 +47,35 @@ public WithWarnings prepend(ArrayRope<Warning> newWarnings) {
+        Warning[] valuesWarnings = warningsLibrary.getWarnings(value, null);
+        allWarnings = new Warning[valuesWarnings.length + warningsArr.length];
+        System.arraycopy(warningsArr, 0, allWarnings, 0, warningsArr.length);
+        System.arraycopy(valuesWarnings, 0, allWarnings, warningsArr.length, valuesWarnings.length);
```
Here we are trying hard to create `Warning[]` and then we are going to convert it back to `TruffleObject` I assume.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/error/Warning.java:
```diff
@@ -104,8 +104,16 @@ public static Array getAll(WithWarnings value) {
+  public static Array getAll(Object value, WarningsLibrary warnings) {
+    if (warnings.hasWarnings(value)) {
+      try {
+        return new Array((Object[]) warnings.getWarnings(value, null));
```
Here we take the array and wrap it again into `TruffleObject`. Maybe `ArrayRope` itself could be a `TruffleObject`?

hubertp replied:
My main concern is that we have a contract on returning Warnings in a specific order. I think it is a topic worth exploring but potentially in a separate PR.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java:
```diff
@@ -195,17 +197,21 @@ Object doPolyglot(
         profiles[i].enter();
         return r;
-      } else if (r instanceof WithWarnings w) {
+      } else if (warnings.hasWarnings(r)) {
```
This is certainly better coding style. We are close to fixing multi-dimensional arrays, I think. Can you test invocation of a Java method that accepts `long[][]` and there is a `Number` with warning somewhere deep in the 2nd level Enso array?

hubertp replied:
Yeah, the current approach doesn't work with multi-dimensional arrays, yet. I will see if I can do anything about it.

hubertp replied:
Now we do.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/text/util/ExpectStringNode.java:
```diff
@@ -31,6 +32,11 @@ String doString(String str) {
   }
 
+  @Specialization
+  String doWarning(WithWarnings value) {
```
Is this correct after introduction of `WarningsLibrary`? Shouldn't one guard `WarningsLibrary.hasWarnings` and then `WarningsLibrary.removeWarnings(value)`?

hubertp replied:
Yeah, this needs an update as well.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/Array.java:
```diff
@@ -128,6 +131,54 @@ boolean hasType() {
     return true;
   }
 
+  @ExplodeLoop
```
Explode loop is unlikely to work as `items` isn't a compile time constant. You'd have to use a `ValueProfile.profile(items.length)` or co., but I am not sure how often one processes arrays of the same size? Maybe vector literals?

hubertp replied:
Yeah, Vector literals. But I doubt it is worth it at that point, thanks.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/Array.java:
```diff
@@ -128,6 +131,54 @@ boolean hasType() {
+  @ExportMessage
+  boolean hasWarnings() {
+    if (withWarnings == null) {
+      withWarnings = hasWarningElements(items);
```
How it can be `null` when the same `hasWarningsElements` function is invoke in the constructor?

hubertp replied:
Survived some refactoring. Fixed

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/Array.java:
```diff
@@ -128,6 +131,54 @@ boolean hasType() {
+  Warning[] getWarnings(Node location) {
+    ArrayRope<Warning> warnings = new ArrayRope<>();
+    for (int i = 0; i < items.length; i++) {
+      if (items[i] instanceof WithWarnings w) {
```
Shouldn't this use `WarningsLibrary`? `@ExplodeLoop` is wrong here.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/Array.java:
```diff
@@ -128,6 +131,54 @@ boolean hasType() {
+  Array removeWarnings() {
+    Object[] items = new Object[this.items.length];
+    for (int i = 0; i < this.items.length; i++) {
+      if (this.items[i] instanceof WithWarnings w) {
```
Same comment as for `getWarnings` - we probably need to use the `WithWarnings` library to handle multi-dimensional arrays.

JaroslavTulach reviewed (approved):
This is certainly step towards better implementation of `WithWarnings`. I left few questions inline - mostly related to multi-dimensional arrays.

hubertp commented:
Checked the benchmarks and they look OK

hubertp committed:
Support multi-dimensional Vectors with warnings

radeusgd commented on test/Tests/src/Semantic/Warnings_Spec.enso:
```diff
@@ -153,9 +157,55 @@ spec = Test.group "Dataflow Warnings" <|
+        even x = (Warning.attach x x) % 2
+        res = [1,2,3,4].map even
+
+        res . should_equal [1, 0, 1, 0]
+        Warning.get_all res . map .value . should_equal [4, 3, 2, 1]
+        res.at 0 . should_equal 1
+        Warning.get_all (res.at 0) . map .value . should_equal [1]
```
Tbh I don't understand this logic.

How if `res` has warnings `[4, 3, 2, 1]`, the call to `res.at 0` has only `[1]`?

To be consistent with the rest of our logic I think the mere fact that we are calling _something_ on `res` means the result should get 'infected' with _all_ warnings of `res`.

I know that in case of arrays it may seem more intuitive to just return the warnings associated with the stored value. But this is inconsistent with other of our primitives like atoms - see for example:
![image](https://user-images.githubusercontent.com/1436948/206010993-935c5cfe-17a8-4508-a433-27c02fe0c335.png)

radeusgd replied:
In reality, it actually depends on the actual _meaning_ of a given warning whether it should infect the whole collection or just keep itself to a single element and only show up in whole collection accesses.

Still, if we are going for 'stick warnings with elements instead of whole array', what about the test below?

```diff
         slice = res.slice 1 4
-        Warning.get_all slice . map .value . should_equal [4, 3, 2, 1]
+        Warning.get_all slice . map .value . should_equal [4, 3, 2]
```

IMO we should change it to the following to reflect that semantics. Of course this complicates slicing quite a lot...

I just want to demonstrate that if we want this to behave in a consistent way, making `res.at 0` return _just_ warnings of the first element has some complicated implications.

radeusgd replied:
Finally, I'm not against strongly against `res.at 0` returning _just_ `[1]`. But then, shouldn't in my REPL example above, `p.first` also only return `["a"]`?

radeusgd replied:
Just checked on your branch to verify the behaviour I shown in the REPL did not change in this PR - but it is the same as above.

![image](https://user-images.githubusercontent.com/1436948/206013597-1e28c97c-a6ed-4bc5-8ae1-7aa9b8560bf1.png)

hubertp replied:
The reason why all this propagation of warnings happens for fields/methods is due to the invocation logic that explicitly extracts/propagates warnings for `self` arguments.
The reason why it doesn't work that way for Vector (and anything Array-like) is because accessing elements is completely done via interop. So it was on purpose. 

I was going to delay this until we have a separate discussion but since you brought it up now I picked up the slack. This is actually fairly to easy to implement now with `WarningsLibrary` as demonstrated in the last commit, although the whole warnings stuff pollutes a lot of the logic :shrug:

radeusgd replied:
Thanks!

hubertp committed:
Propagate warnings from case branches

hubertp committed:
nit

hubertp committed:
Propagate all vector warnings when reading element

Previously, accessing an element of an Array-like structure would only
return warnings of that element or of the structure itself.
Now, accessing an element also returns warnings from all its elements as
well.
